### PR TITLE
Separate go module for tools to avoid polluting root go.mod

### PR DESCRIPTION
**Pull request by dims**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Now that we have GOPROXY, Do we really want to vendor stuff? new repo-infra drags in a bunch of stuff (see the comments from @liggitt in klog v2 PR - https://github.com/kubernetes/kubernetes/pull/90183#issuecomment-615278050)

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
> Now that we have GOPROXY, Do we really want to vendor stuff? new repo-infra drags in a bunch of stuff (see the klog v2 PR - #90183)

yeah, FWIW in kind the approach is a step further:
- `hack/tools/go.mod` exists, this makes `hack/tools` a distinct module from `sigs.k8s.io/kind/go.mod`
- `hack/tools/tools.go` undescore imports development tools / linters
- `hack/*.sh` scripts cd to `hack/tools` and `go install` tools from there

this lets us:
- not keep tools in the root go.mod / go.sum
- not require end users just doing a binary build to download development dependencies
- not require development dependencies to align with binary dependencies
  - gives us a better idea what dependencies we actually depend on in code we ship
- can still be fully cached & reproducible (go.sum, go module cache, go build cache)
- not checking in vendor makes cloning much faster and avoids bloating the repo over time

k/k is already ~ 800MB just for `.git` last I checked (!)

**Comment by liggitt** [comment]
I like the idea of isolated tools... still lets us pin versions, still lets us do a `cd tools && go mod download` to fetch all deps locally for offline dev if needed

**Comment by dims** [comment]
once we establish this pattern for bazel stuff, we can start moving other tools from build/tools.go one by one.

File: hack/tools/go.mod [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module sigs.k8s.io/kind/hack/tools
```
Comment (liggitt) [inline hack/tools/go.mod:1]: copy/paste :)

File: hack/tools/go.mod [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+go 1.13
+
+require (
+	github.com/bazelbuild/bazel-gazelle v0.19.1-0.20191105222053-70208cbdc798
+	github.com/bazelbuild/buildtools v0.0.0-20190917191645-69366ca98f89
+	k8s.io/repo-infra v0.0.1-alpha.1
+)
```
Comment (liggitt) [inline hack/tools/go.mod:9]: was this tidied with `go mod tidy`? Does anything make sure that is done before PRs merge?

File: hack/tools/go.mod (hunk 1 above)
Comment (dims) [reply hack/tools/go.mod:1]: LOL yep

File: hack/tools/go.mod (hunk 2 above)
Comment (dims) [reply hack/tools/go.mod:9]: the verify script runs hack/update-bazel ... which fetches the dependencies ... which ends up updating go.sum. So the verify job may catch a dirty go.sum? (need to confirm)

**Comment by BenTheElder** [comment]
We should probably add something to verify to do go mod tidy and diff on this new module.

File: hack/tools/go.mod (hunk 2 above)
Comment (BenTheElder) [reply hack/tools/go.mod:9]: That wouldn't tidy though?

File: Godeps/LICENSES [hunk 3]
```diff
@@ -2542,237 +2542,6 @@ THE SOFTWARE.
 ================================================================================
 
 
-================================================================================
-= vendor/github.com/bazelbuild/bazel-gazelle licensed under: =
```
Comment (BenTheElder) [inline Godeps/LICENSES:2546]: I suppose we may need to be more diligent about licenses for things under hack/tools now?

Comment (liggitt) [reply Godeps/LICENSES:2546]: hmm… might be good to get an ack on what license requirements are for code we don't ship or link to, but make use of in dev tools/scripts. paging @kubernetes/licensing ...

File: hack/tools/go.mod (hunk 2 above)
Comment (dims) [reply hack/tools/go.mod:9]: i'd vote for adding update/verify later. honestly these don't change quickly and one of us can help for now.

File: Godeps/LICENSES (hunk 3 above)
Comment (dims) [reply Godeps/LICENSES:2546]: right

Comment (dims) [reply Godeps/LICENSES:2546]: will have to file a issue to follow up on this

File: hack/tools/go.mod (hunk 2 above)
Comment (dims) [reply hack/tools/go.mod:9]: (and tidy)

File: Godeps/LICENSES (hunk 3 above)
Comment (BenTheElder) [reply Godeps/LICENSES:2546]: i meant we'll have lower visibility into e.g. changes, reviewers need to actually make a point to check these versus just looking at this file

Comment (BenTheElder) [reply Godeps/LICENSES:2546]: @kubernetes/dep-approvers

Comment (dims) [reply Godeps/LICENSES:2546]: @BenTheElder yep

File: hack/tools/go.mod (hunk 2 above)
Comment (dims) [reply hack/tools/go.mod:14]: added update/verify scripts.

File: Godeps/LICENSES (hunk 3 above)
Comment (dims) [reply Godeps/LICENSES:2546]: added support for a separate LICENSES file under hack/tools (which is updated by the update script and verified by the verify script)

**Comment by dims** [comment]
```
$ rg hack-tool build-log.txt
71379:Verifying verify-hack-tools.sh
71381:+++ Running case: verify.hack-tools
71383:+++ command: bash "hack/make-rules/../../hack/verify-hack-tools.sh"
71769:SUCCESS  verify-hack-tools.sh	15s
```

File: hack/update-bazel.sh [hunk 4]
```diff
@@ -104,3 +103,6 @@ buildozer -quiet 'set clinkopts select({"@io_bazel_rules_go//go/platform:linux":
 if [[ $ret != 0 && $ret != 3 ]]; then
   exit 1
 fi
+
+# Avoid bazel stuff in tools/ directory
+rm hack/tools/BUILD
```
Comment (liggitt) [inline hack/update-bazel.sh:108]: is this still needed with the exclusion of hack/tools in `build/root/.kazelcfg.json`?

File: hack/make-rules/test.sh [hunk 5]
```diff
@@ -226,7 +226,7 @@ produceJUnitXMLReport() {
 
   if ! command -v gotestsum >/dev/null 2>&1; then
     kube::log::error "gotestsum not found; please install with " \
-      "GO111MODULE=off go install k8s.io/kubernetes/vendor/gotest.tools/gotestsum"
+      "GO111MODULE=off go install gotest.tools/gotestsum"
```
Comment (liggitt) [inline hack/make-rules/test.sh:229]: doesn't this need `GO111MODULE=on`?

File: go.mod [hunk 6]
```diff
@@ -72,7 +70,6 @@ require (
 	github.com/lithammer/dedent v1.1.0
 	github.com/lpabon/godbc v0.1.1 // indirect
 	github.com/magiconair/properties v1.8.1 // indirect
-	github.com/mattn/go-isatty v0.0.9 // indirect
```
Comment (liggitt) [inline go.mod]: dropping this is what required the drop to v0.0.4 in a later commit

File: hack/update-hack-tools.sh [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+
+# Explicitly opt into go modules, even though we're inside a GOPATH directory
+export GO111MODULE=on
+
+# Detect problematic GOPROXY settings that prevent lookup of dependencies
+if [[ "${GOPROXY:-}" == "off" ]]; then
+  kube::log::error "Cannot run hack/update-vendor.sh with \$GOPROXY=off"
```
Comment (liggitt) [inline hack/update-hack-tools.sh:29]: nit: wrong script name

File: hack/update-hack-tools.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  exit 1
+fi
+
+kube::golang::verify_go_version
+
+pushd "${KUBE_ROOT}/hack/tools" >/dev/null
+  echo "=== tidying go.mod/go.sum in hack/tools"
+  go mod edit -fmt
+  go mod download
```
Comment (liggitt) [inline hack/update-hack-tools.sh]: what does the `go mod download` accomplish?

File: hack/update-hack-tools.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  go mod tidy
+
+  if [[ "${CALLED_FROM_VERIFY:-""}" == "1" ]]; then
+    go clean -cache -modcache -r
```
Comment (liggitt) [inline hack/update-hack-tools.sh:45]: I didn't expect force clearing the modcache... why was this needed?

File: hack/update-hack-tools.sh [hunk 10]
```diff
@@ -35,9 +35,12 @@ kube::golang::verify_go_version
 pushd "${KUBE_ROOT}/hack/tools" >/dev/null
   echo "=== tidying go.mod/go.sum in hack/tools"
   go mod edit -fmt
-  go mod download
+  go mod vendor
   go mod tidy
```
Comment (liggitt) [inline hack/update-hack-tools.sh:39]: not sure it matters, but tidy before vendor?

File: hack/tools/Makefile [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+	$(GOBIN)/staticcheck\
+
+.PHONY: all
+all: clean build
+
+.PHONY: clean
+clean:
+	rm -f $(DEP_PROGS)
+	rm -rf vendor/
+	go clean
```
Comment (liggitt) [inline hack/tools/Makefile:35]: I'm not familiar with what this does with no args

File: hack/update-bazel.sh (hunk 4 above)
Comment (dims) [reply hack/update-bazel.sh:108]: yes, apparently the exclusion is not enough.

File: hack/make-rules/test.sh (hunk 5 above)
Comment (dims) [reply hack/make-rules/test.sh:229]: Done!

File: go.mod (hunk 6 above)
Comment (dims) [reply go.mod]: yes.

File: hack/update-hack-tools.sh (hunk 7 above)
Comment (dims) [reply hack/update-hack-tools.sh:29]: fixed

**Comment by liggitt** [comment]
just a few questions/nits

File: hack/update-hack-tools.sh (hunk 8 above)
Comment (dims) [reply hack/update-hack-tools.sh]: yea, i cleaned it up in the next commit

File: hack/update-hack-tools.sh (hunk 9 above)
Comment (dims) [reply hack/update-hack-tools.sh:45]: in the verify-hack-tools script, we create a temp directory and use that for the go path etc, i was unable to clean it up unless we cleared the modcache because of https://github.com/golang/go/issues/27161

File: hack/update-hack-tools.sh (hunk 10 above)
Comment (dims) [reply hack/update-hack-tools.sh:39]: ack switching it around

File: hack/tools/Makefile (hunk 11 above)
Comment (dims) [reply hack/tools/Makefile:35]: i picked it from somewhere. dropping it

File: hack/update-hack-tools.sh (hunk 9 above)
Comment (liggitt) [reply hack/update-hack-tools.sh:45]: if everything in update-hack-tools uses gomodule mode, you shouldn't need to set a fake GOPATH

Comment (liggitt) [reply hack/update-hack-tools.sh:45]: that means you can just copy the hack/tools dir (don't need the k8s.io/kubernetes parent dir structure, and can stop overriding GOPATH

Comment (dims) [reply hack/update-hack-tools.sh:45]: good point, let me try that

File: hack/tools/Makefile [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+.PHONY: clean
+clean:
+	rm -f $(DEP_PROGS)
+	rm -rf vendor/
+
+.PHONY: build
+build: deps
+
+.PHONY: deps
+deps: $(DEP_PROGS)
```
Comment (liggitt) [inline hack/tools/Makefile:40]: /hold

seeing a circular dependency error here:
`make: Circular /Users/liggitt/go/bin/buildozer <- deps dependency dropped.`

**Comment by liggitt** [comment]
might want to squash in review comment commits as well

File: hack/tools/Makefile (hunk 12 above)
Comment (dims) [reply hack/tools/Makefile:40]: Fixed.

**Comment by cblecker** [comment]
This is great. Thank you all for this <3

**Comment by BenTheElder** [comment]
Excited to see this merged :-)

File: hack/tools/Makefile [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+export GO111MODULE=on
+
+GOBIN ?= $(shell go env GOPATH)/bin
```
Comment (BenTheElder) [inline hack/tools/Makefile:17]: probably we should prefer a local binary dir instead of clobbering the system binaries?

Comment (dims) [reply hack/tools/Makefile:17]: We don't use this from our build harness it can be used by folks who want to work offline, we could make it clearer in the README that they can override the GOBIN to place it elsewhere

Comment (BenTheElder) [reply hack/tools/Makefile:17]: ack
